modules/mono/editor/GodotTools/GodotTools.ProjectEditor/ProjectUtils.cs

```diff
@@ -59,5 +67,99 @@ public static void EnsureGodotSdkIsUpToDate(MSBuildProject project)
             root.Sdk = godotSdkAttrValue;
             project.HasUnsavedChanges = true;
         }
+
+        public static void EnsureTargetFrameworkIsSet(MSBuildProject project)
+        {
+            bool HasNoCondition(ProjectElement element) => string.IsNullOrEmpty(element.Condition);
+
+            bool ConditionMatches(ProjectElement element, string platform) =>
+                string.Compare(element.Condition.Trim().Replace(" ", ""),
+                    $"'$(GodotTargetPlatform)'=='{platform}'", StringComparison.OrdinalIgnoreCase) == 0;
+
+            // if the existing framework is equal or higher than ours, we're good
+            bool IsVersionUsable(string theirs, string ours) =>
+                Version.TryParse(theirs.Substring(3), out var versionTheirs) &&
+                Version.TryParse(ours.Substring(3), out var versionOurs) &&
+                versionTheirs >= versionOurs;
+
+            // if the property already has our condition, we're good
+            // if the property has no conditions and the group has no conditions, we can use it
+            // if the property has no conditions but the group matches our condition, cool
+            bool IsFrameworkUsable(ProjectPropertyGroupElement group, ProjectPropertyElement prop,
+                string platform, string framework)
+            {
+                return IsVersionUsable(prop.Value, framework) &&
+                       (ConditionMatches(prop, platform) ||
+                        (HasNoCondition(prop) &&
+                         (HasNoCondition(group) || ConditionMatches(group, platform))));
+            }
+
+            // If the condition matches (in the property or group) but the target framework isn't high enough, replace it
+            bool ShouldReplaceProperty(ProjectPropertyGroupElement group, ProjectPropertyElement prop,
+                string platform, string framework)
+            {
+                return !IsVersionUsable(prop.Value, framework) &&
+                       (ConditionMatches(prop, platform) ||
+                        (HasNoCondition(prop) && ConditionMatches(group, platform)));
+            }
+
+            ProjectPropertyGroupElement mainGroup = null;
+            Dictionary<ProjectPropertyGroupElement, List<ProjectPropertyElement>> propertiesToRemove = new();
+
+            var root = project.Root;
+            bool skipProjectUpdate = false;
+
+            foreach (var group in root.PropertyGroups)
+            {
+                foreach (var prop in group.Properties)
+                {
+                    // If the user sets the GodotSkipProjectUpdate property, we don't do anything
+                    if (string.Compare(prop.Name, "GodotSkipProjectUpdate", StringComparison.OrdinalIgnoreCase) == 0)
+                    {
+                        skipProjectUpdate = prop.Value?.ToLower() == "true" || prop.Value == "1";
+                    }
```

Comment:
I need this 😄  I've renamed it to something longer and more explicit of what it means.

Reply:
I'm still a bit hesitant to adding this property. Can you describe your use case? I'm also not sure if it should be an MSBuild property or a ProjectSetting/EditorSetting, it feels weird for it to be an MSBuild property to be honest since it doesn't affect the build.

I was hoping we could merge this for 4.4.dev5, and this is the only blocker for me, so I was wondering if it'd be acceptable to you to extract this part to a separate PR so we can give it more time to discuss, without delaying the rest of the PR.

Reply:
I don't understand what the problem is of having a property that allows an advanced user to bypass the automatic rewriting of the csproj? I'm having to maintain custom .net SDKs for consoles, and the automatic rewriting is not great when I'm testing or when I'm needing to change the settings for specific use cases. Why would having this option be a blocker for you? It is not documented, very obvious about what it does, and it's only for advanced use cases. I would really like to not have to maintain custom patches to be able to bypass certain things, it causes a ton of conflicts every time we merge upstream godot into our forks.

Reply:
You certainly have your point; however, shouldn't the necessity for this feature be discussed and implemented in a separate PR? Even if this feature is not merged upstream, you can keep it in your local fork to enhance your merging experience.

Reply:
Keeping it in my local fork is precisely the thing that causes no end of conflicts whenever we merge upstream to our fork, and precisely what I want to avoid.

Reply:
Every feature that gets added to the engine needs to be thoroughly discussed. And being undocumented won't prevent it from being used, which means users may start relying on it and complain later if we want to rename it, move it somewhere else, etc.

The rewriting of the csproj would happen anyway even with `GodotSkipAutomaticTargetFrameworkUpdate` because Godot will change the `Sdk` to `Godot.NET.Sdk/{Version}`. Does this not affect you?

Also, I don't understand why this would affect consoles if this PR was originally only supposed to update the TFM for Android and iOS platforms. And the TFM shouldn't change if it satisfies the minimum requirements anyway, are you saying consoles are using a .NET version older than the minimum requirement set by Godot?

Reply:
Given that you mentioned above that you would like to enforce the minimum required version for all platforms, then yes that affects all platforms, not just consoles. Rewriting the godot sdk net version doesn't affect consoles as much as this, and yes, consoles can follow a different version than the rest of the platforms. And given that the current target framework version is 6, not 8, and that this PR is on top of the current 4.3 head, not on top of your net8 uodate, yes, different platforms will have different requirements, the consoles will need to have uodated requirements on these (a local patch I'm not keen on maintaining but can't go around that one), and having the ability to control whether these properties get automatically changed is important.

This is an integral part of this PR, it is not an optional thing to be discussed in a committee somewhere. If you prefer it documented, I will gladly document it. If you refuse to take this as part of this PR, I will withdraw this PR entirely and the consent for using it.

Reply:
I understand that console platforms may have different requirements. In 4.3, the minimum required TFM is `net6.0` but Android requires `net7.0` and iOS requires `net8.0`. To handle this we add additional `TargetFramework` properties with a condition to match the `GodotTargetPlatform`, I don't see why consoles can't do the same.

This PR would not touch those additional `TargetFramework` properties, it should only affect the ones that have no conditions or the ones that have a condition that matches the pattern `$(GodotTargetPlatform)='{platform}'` where `{platform}` is one of the hardcoded platforms (currently only `android` and `ios`).